### PR TITLE
WIP: Add cronjob timezone

### LINKED ISSUE #47202: CronJobs should support timezones
I couldn't find any discussion of this after searching for "cron timezone", "cronjob timezone", or "scheduledjob timezone".

The CronJob spec makes reference to https://en.wikipedia.org/wiki/Cron. That page suggests that cron would respect the timezone for a given user. The controller manager runs in a single time zone under a single user so I can't use different time zones for each job. I have jobs that run based on the schedule of external entities that observe daylight savings time. So, if I define that CronJob in UTC I will be forced to update that job from time to time (generally not something one remembers to do after just losing an hour of sleep).

I see two options for how this support might work in kubernetes:

1. Add some new field to the `CronJobSpec`, like `timezone: "Americas/Chicago"`.
2. Use an extended cron syntax that includes timezone, e.g. `30 6 * * 1 Europe/Stockholm`

### PR DESCRIPTION
**Pull request by iterion**
#### What type of PR is this?

/kind feature
/kind api-change

#### What this PR does / why we need it:

This PR adds a new field to the `CronJobSpec` named `TimeZone` that allows users to choose the specific time zone to use when scheduling jobs. If not specified, the previous default behavior of using the local time for the `kube-controller-manager` is preserved.

The ability to specify time zones is useful for Kubernetes users in situations where application requirements necessitate running jobs in specific time zones. In situations where users don't have control over the control plane local time, this also allows users to choose a time zone that is consistent and known to them.

#### Which issue(s) this PR fixes:

Fixes #47202

#### Does this PR introduce a user-facing change?

```release-note
Added the ability to specify TimeZone as part of the CronJobSpec. 
This allows the end-user to specify the time zone to be used when determining when to run a CronJob. If not specified, the time zone will continue to default to the same used by the kube-controller-manager.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-apps/19-Graduate-CronJob-to-Stable#support-timezone-for-cronjobs
```

### CODE REVIEW

**Comment by fedebongio** [comment]
/remove-sig api-machinery
(looks like it's sig apps)

File: pkg/apis/batch/types.go [hunk 1]
```diff
@@ -359,6 +359,11 @@ type CronJobSpec struct {
 	// This is a pointer to distinguish between explicit zero and not specified.
 	// +optional
 	FailedJobsHistoryLimit *int32
+
+	// The time zone to be used when scheduling the job given the schedule specified above.
+	// Valid options include anything in the IANA Time Zone database.
+	// +optional
+	TimeZone string
```
Comment (dims) [inline pkg/apis/batch/types.go:366]: should be `*string` for it to be optional

File: staging/src/k8s.io/api/batch/v1beta1/types.go [hunk 2]
```diff
@@ -136,6 +136,11 @@ type CronJobSpec struct {
 	// Defaults to 1.
 	// +optional
 	FailedJobsHistoryLimit *int32 `json:"failedJobsHistoryLimit,omitempty" protobuf:"varint,7,opt,name=failedJobsHistoryLimit"`
+
+	// The time zone to be used when scheduling the job given the schedule specified above.
+	// Valid options include anything in the IANA Time Zone database.
+	// +optional
+	TimeZone string `json:"timeZone,omitempty" protobuf:"bytes,8,opt,name=timeZone"`
```
Comment (dims) [inline staging/src/k8s.io/api/batch/v1beta1/types.go:143]: same

**Comment by dims** [comment]
cc @soltysh @alaypatel07

**Comment by soltysh** [comment]
/hold
Thank you very much for your contribution, but before accepting this we'll need to revisit the decision stated in https://github.com/kubernetes/kubernetes/issues/47202#issuecomment-360820586 We need full support both from SIG-Apps and SIG-Architecture followed by KEP to be able to accept it.

**Comment by iterion** [comment]
> Instead of putting this in Kubernetes the ask is to:
> * Develop this in the ecosystem (e.g., a controller) that others can use. Distribute it, solve the problems there, and see what update looks like
> * If the solution is widely adopted and can be used by everyone (including small scale, multi-cluster, etc) then it could be considered for core Kubernetes

It seems like the ask has been satisfied, but maybe we need clearer criteria? How do we move this forward?

I'm happy to help out with a KEP, and to advocate for this, but I'm not sure where to start.

**Comment by soltysh** [comment]
> It seems like the ask has been satisfied, but maybe we need clearer criteria? How do we move this forward?
> 
> I'm happy to help out with a KEP, and to advocate for this, but I'm not sure where to start.

Start with presenting the idea to sig-apps at one of our next meetings. We can make the initial decision there and we can get the final approval from sig-architecture after that. If you're willing to write down the KEP we can easily plan this for 1.23.

**Comment by soltysh** [comment]
Feel free to ping me on slack and I can get you pointed and setup where needed.

**Comment by soltysh** [comment]
@iterion do you think you could make the next sig-apps call on July 26th to talk about this?

**Comment by iterion** [comment]
@soltysh sorry for the delay. I should be able to make it on July 26, what time does the meeting happen? My calendar is open for much of the day.

**Comment by soltysh** [comment]
> @soltysh sorry for the delay. I should be able to make it on July 26, what time does the meeting happen? My calendar is open for much of the day.

@iterion it's [9am pacific](http://www.thetimezoneconverter.com/?t=9:00&tz=PT%20%28Pacific%20Time%29)

**Comment by iterion** [comment]
Cool, I'm available to attend.

**Comment by mtparet** [comment]
@iterion Thank you so much. Most of k8s users are waiting this feature, thanks again caring about us !

**Comment by cndoit18** [comment]
Hello, the current cronjob already supports timezone handling.
see: https://github.com/kubernetes/kubernetes/pull/104404

**Comment by nitaigao** [comment]
The previous method using `CRON_TZ=` is being flagged as unsupported, can this be reopened?

**Comment by cbandy** [comment]
See #47202

**Pull request closed without merging**